node: v25

Renegade334 opened the pull request:
#### Breaking API changes
- **assert:** move CallTracker to EOL
- **assert:** move assert.fail with multiple arguments to eol
- **buffer:** move SlowBuffer to EOL
- **fs:** move rmdir recursive option to end-of-life
- **perf_hooks:** move deprecated accessors to EOF
- **process:** move multipleResolves event to EOL

#### New API features
- **lib:** expose global ErrorEvent
- **src:** add an option to make compile cache portable
- **util,console:** colorize regexp groups, character classes, etc
- **v8:** add cpu profile
- The experimental `node:quic` module now has type definitions.

#### DOM types
- All of the DOM interfaces have had their definitions synchronised with the TS DOM libraries (with Node.js-specific bolt-ons where applicable). This should improve interoperability, particularly with the streams interfaces where handling and persistence of the type parameter were previously quite divergent.
- Added event mapping behaviour to EventTargets, aligning with lib.dom.

#### Breaking/observable type changes
- `node:`-qualified modules are canonicalised. Barenamed builtin modules are aliased to the qualified module declaration. This is observable if using the TS API.
- The remaining subpathed modules whose definitions were "squashed" into the root module (`node:util/types` etc.) have been spun out into separate .d.ts files. This is observable if using the TS API.
- The fake ESM-with-default-export structure of `node:cluster` has been fixed. Top-level imports should now be correctly transpiled.
- Some deprecated symbols have been removed.
- EventEmitter changes from #73882. EventEmitter no longer accepts `Function` objects as arguments, requiring proper function-type arguments instead. The EventEmitter type parameter default has been changed to `any`.
- Key generation/exporting in `node:crypto` has been completely overhauled, as quite a bit of its behaviour was incorrect and the definitions were growing rapidly in complexity with the new crypto standards. The new type system is more robust and cuts the module size by around 20%, but some of the changes to helper types are not backwards-compatible if imported directly in userland.
- The type behaviour of the custom inspect formatting features has had to be significantly altered to reflect the fact that "colors" can now be callbacks. This is not fully backwards-compatible for certain patterns of modifying the customisation namespace.

Renegade334 committed:
crypto: refactor key export behaviour

Renegade334 committed:
events: overhaul for v25

Co-authored-by: ckohen <chaikohen@gmail.com>

Renegade334 committed:
events: add InternalEventEmitter

Renegade334 committed:
events: remove emit tests for internal emitters

`emit()` on internal emitters accepts any input, so these aren't
testing anything at all, as evidenced by the fact that a lot of these
are passing supposedly invalid input without errors.

Renegade334 committed:
assert: move CallTracker to EOL

Renegade334 committed:
assert: move assert.fail with multiple arguments to eol

Renegade334 committed:
buffer: move SlowBuffer to EOL

Renegade334 committed:
meta: add "node:quic" module

jonasgeiler committed:
worker_threads: use `unknown` for worker error event

Renegade334 committed:
fs: move rmdir recursive option to end-of-life

Renegade334 committed:
perf_hooks: move deprecated accessors to EOF

Renegade334 committed:
process: move multipleResolves event to EOL

Renegade334 committed:
lib: expose global ErrorEvent

Renegade334 committed:
util,console: colorize regexp groups, character classes, etc

Renegade334 committed:
src: add an option to make compile cache portable

Renegade334 committed:
v8: add cpu profile

Renegade334 committed:
meta: canonicalise "node:" module qualifiers

Renegade334 committed:
meta: add NodeJS global buffer source types

Renegade334 committed:
buffer: DOM types

Renegade334 committed:
console: DOM types

Renegade334 committed:
crypto: DOM types

Renegade334 committed:
meta: IDE warnings for compatibility types

Renegade334 committed:
perf_hooks: DOM types

Renegade334 committed:
module: DOM types

Renegade334 committed:
timers: DOM types

Renegade334 committed:
url: DOM types

Renegade334 committed:
util: DOM types

Renegade334 committed:
worker_threads: DOM types

Renegade334 committed:
stream/web: DOM types

Renegade334 committed:
cluster: remove faux default export

Renegade334 committed:
meta: branch to v25

Renegade334 committed:
doc: `domain.add()` does not accept timer objects

Renegade334 committed:
meta: remove legacy symbols

y-hsgw committed:
process: change `disconnect()` to optional `disconnect?()`

Renegade334 committed:
events: add EventTarget event maps

Renegade334 commented:
Since this is going to be a PITA to review, here are some chunked diffs:

- dc2653694eb5f6cf6630cdbf93243ccdee0e7358...e4f508b9b32c2fd38d3003a75f79a0a0df60b439 are the run-of-the-mill definition changes.
- 9969d9c747138d3a611ff6f6c5b49a8bc128d2b1 makes `node:` the canonical module schema.
- 83533e075976dab35f6a548e37c11b639345e170...ce9402069bc1e45386f29e78d27c730d466fc26b are the changes syncing the web types with lib.dom.
- db34466089fcddb0145f758ddc062acd3311154e makes `node:cluster` a proper CJS module with a top-level export, as opposed to a fake ESM module with a pseudo-default export.
- e608186d01070857c4f74bf104896c36423a8ebf gets rid of some deprecated exports.
- e3d3c04cd03505af097196676d3fb388b9544eb2 moves the key generation/export methods in `node:crypto` away from the fairly disgusting overload blancmange (which didn't work as intended in a lot of cases) to more robust behaviours using helper types.
- f2655033eaf04cd3ddf89b4ce47a383a5024c5b6...4c2cc8a1f2bc8c03dd082626ab1341cf4cec1b1f are the events changes as previously advertised.

jakebailey reviewed (commented):
Read through everything; overall seems good, but had some comments.

Renegade334 commented on types/node/test/process.ts:
```diff
@@ -33,10 +33,8 @@ import { fileURLToPath } from "node:url";
         (message: any, sendHandle: net.Server | net.Socket | dgram.Socket | undefined) => {},
     );
     process.prependOnceListener("SIGBREAK", () => {});
-    process.emit("SIGINT");
```
It's not a break, it's just that the `emit()` signatures for internal event emitters are (and were) fully permissive, so these "are the args valid for this event type" tests are pointless. Although the overloads provide type _hinting_, the catch-all signature will (by design) allow absolutely any parameter combo to be passed here – and many of the tests were themselves passing "invalid" args, as if to demonstrate the point! I got rid of the whole lot in 51ed7ef6dda2fde5fbf9feae9da57876c95fb5fc accordingly.

Renegade334 committed:
process: deprecate global listener types

These are no longer necessary internally, but are used in some
fairly high-exposure libraries - probably best not to remove
them entirely.

Renegade334 commented on types/mz/fs.d.ts:
```diff
@@ -574,7 +573,7 @@ export function rmdir(path: PathLike, callback: NoParamCallback): void;
  * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
  */
-export function rmdir(path: PathLike, options: RmDirOptions, callback: NoParamCallback): void;
+export function rmdir(path: PathLike, options: object, callback: NoParamCallback): void;
```
As an empty interface?

jakebailey replied:
That's probably ok, though I was thinking just keeping it as is and deprecating it. But I'm pretty conservative about breaks 😅

Renegade334 committed:
fs: export empty RmDirOptions

jakebailey reviewed (commented):
I think this is now ready to go. To note it, I asked the team about adding an odd-numbered version and people did not express any concerns about that, so long as you're prepared to futz with more dirs until (if?) Node switches away from even-number-stable versioning.

I think the EventEmitter stuff in its current form should turn out fine, though if it turns out to be breaky we probably should have a revert ready for that specific thing?

My final question is about the movement of the canonical definition from say `fs` to `node:fs`. If people are augmenting `fs`, will that now no longer work if `node:fs` is the one that contains the declarations?

Renegade334 committed:
stream: move legacy interfaces to stream.d.ts

Renegade334 committed:
stream: general sort-out

- improvements/corrections to pipeline behaviour
- move readable.compose
- add stream.compose
- correct use of NodeJS.*Stream and web.*Stream
- use aliased promisifiers
- a few assorted type corrections

Renegade334 commented on types/node/stream.d.ts:
```diff
@@ -1333,46 +1367,57 @@ declare module "node:stream" {
+            err: NodeJS.ErrnoException | null,
+            value: S extends (...args: any[]) => PromiseLike<infer R> ? R : undefined,
+        ) => void;
+        type PipelineResult<S extends PipelineDestination<any, any>> = S extends NodeJS.WritableStream ? S : Duplex;
```
These changes lend some advantages:
- the return values of generator-style transformers are passed through the pipeline as the original type, without narrowing to `AsyncIterator`
- web streams are permitted, and chunk types are inferred from web streams
- non-stream destinations result correctly in a Duplex

All of the existing tests pass, modulo some intentional ExpectTypes.

Renegade334 commented on types/node/stream.d.ts:
```diff
@@ -22,25 +22,26 @@ declare module "node:stream" {
-            stream: T | ComposeFnParam | Iterable<T> | AsyncIterable<T>,
-            options?: { signal: AbortSignal },
+            options?: Stream.PipeOptions,
         ): T;
```
This is definitely not a method on the legacy Stream prototype, and the definition was extremely dodgy. Moved to the Readable prototype and expanded to use the new helper types.

Renegade334 commented:
Unearthed some stashed changes to the stream module at the eleventh hour that were deferred from #71808 for a major version. I think they'd be worth including if you're in the mood to review, but can undo if preferable.

jakebailey commented on types/readable-stream/index.d.ts:
```diff
@@ -430,10 +432,6 @@ declare namespace _Readable {
 
         constructor(options?: ReadableOptions);
         pipe<T extends NodeJS.WritableStream>(destination: T, options?: { end?: boolean | undefined }): T;
-        compose<T extends NodeJS.ReadableStream>(
```
We're losing the generic on this, no? So could be a break?

Renegade334 commented on types/node/stream/promises.d.ts:
```diff
@@ -74,15 +189,21 @@ declare module "node:stream/promises" {
+    ): PipelineResult<B>;
     function pipeline(
-        streams: ReadonlyArray<NodeJS.ReadableStream | NodeJS.WritableStream | NodeJS.ReadWriteStream>,
+        streams: readonly [PipelineSource<any>, ...PipelineTransform<any, any>[], PipelineDestination<any, any>],
```
The array signature is probably fairly unused anyway. I can't imagine under what circumstances you'd pass an array from a variable here, but having a parameter as a genuine ordered tuple type really necessitates a literal, and there's no reason to use a literal instead of passing individual arguments, so the permissive approach might be easier.

jakebailey reviewed (approved):
Well, I guess if this is ready, let's give it a go...
